sound/soc/sof/pm.c

```diff
@@ -229,9 +229,19 @@ static void sof_suspend_streams(struct snd_sof_dev *sdev)
 static int sof_resume(struct device *dev, bool runtime_resume)
 {
 	struct sof_platform_priv *priv = dev_get_drvdata(dev);
-	struct snd_sof_dev *sdev = dev_get_drvdata(&priv->pdev_pcm->dev);
+	struct snd_sof_dev *sdev;
 	int ret = 0;
 
+	if (!priv) {
+		dev_dbg(dev, "%s: priv is NULL", __func__);
+		return 0;
+	}
+	sdev = dev_get_drvdata(&priv->pdev_pcm->dev);
+	if (!sdev) {
+		dev_dbg(dev, "%s: sdev is NULL", __func__);
+		return 0;
```

Comment:
likewise when is this going to happen? if we ever go through sof_probe() in core.c, then sdev will never be NULL.

Reply:
This will happen in our case. If fw fails to load, the platform_device (not pci_device) driver probe will fail. Sequentially, the driver device-driver architecture will set dev->driver_data to NULL automatically. 
If you are interested in the details, you can refer  to ./driver/base/dd.c file.

Reply:
but if the platform_device probe fails, you will not even invoke pm related functions. Please detail the flow, thanks.

Reply:
the PM functions belongs to pci device, not platform device. PCI device driver is registered successfully.

Reply:
see sof-pci-dev.c:

````c
	/* register sof-audio platform driver */
	ret = sof_create_platform_device(priv);
	if (ret) {
		dev_err(dev, "error: failed to create platform device!\n");
		goto release_regions;
	}

>>> SEE HERE, runtime_pm is not enabled !!!

	/* allow runtime_pm */
	pm_runtime_set_autosuspend_delay(dev, SND_SOF_SUSPEND_DELAY_MS);
	pm_runtime_use_autosuspend(dev);

	/*
	 * runtime pm for pci device is "forbidden" by default.
	 * so call pm_runtime_allow() to enable it.
	 */
	pm_runtime_allow(dev);

	/* follow recommendation in pci-driver.c to decrement usage counter */
	pm_runtime_put_noidle(dev);

	return ret;

release_regions:
	pci_release_regions(pci);

	return ret;

````

Reply:
ok, let's play:
````c
int sof_create_platform_device(struct sof_platform_priv *priv)
{
	struct snd_sof_pdata *sof_pdata = priv->sof_pdata;
	struct device *dev = sof_pdata->dev;

	priv->pdev_pcm =
		platform_device_register_data(dev, "sof-audio",
					      PLATFORM_DEVID_NONE,
					      sof_pdata, sizeof(*sof_pdata));
	if (IS_ERR(priv->pdev_pcm)) {
		dev_err(dev, "error: cannot register device sof-audio. Error %ld\n",
			PTR_ERR(priv->pdev_pcm));
		return PTR_ERR(priv->pdev_pcm);
	}

	return 0;
}
````

this will result in sof_probe being called, so how can sof_probe() fail without the error being bubbled all the way to the pci-device?

Reply:
This is handled by device-driver binding methods in the lower level.

Reply:
sorry man, I follow the code and don't see any evidence that your claims hold water. If you want to make your case you've got to be more convincing or prove me wrong.

Reply:
If you register a device, a corresponding driver probe will be called and vice versa. They are async. device is device, driver is driver : )
Image there is a HW platform device, we will register the device, and register the device driver. Even device driver probe failed, the device is also registered successfully. No reason that driver failure will cause device registering failure, right?